src/plugins/vis_augmenter/public/saved_augment_vis/utils/helpers.ts

```diff
@@ -3,14 +3,46 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
-import { getSavedAugmentVisLoader } from '../../services';
-import { ISavedAugmentVis } from '../../types';
+import { get } from 'lodash';
+import { getSavedAugmentVisLoader, getUISettings } from '../../services';
+import { ISavedAugmentVis } from '../types';
 
 /**
  * Create an augment vis saved object given an object that
  * implements the ISavedAugmentVis interface
  */
 export const createAugmentVisSavedObject = async (AugmentVis: ISavedAugmentVis): Promise<any> => {
   const loader = getSavedAugmentVisLoader();
+  const config = getUISettings();
+
+  const isAugmentationEnabled = config.get('visualization:enablePluginAugmentation');
+  if (!isAugmentationEnabled) {
+    // eslint-disable-next-line no-throw-literal
+    throw 'Visualization augmentation is disabled, please enable visualization:enablePluginAugmentation.';
+  }
+  const maxAssociatedCount = config.get('visualization:enablePluginAugmentation.maxPluginObjects');
+
+  await loader.findAll().then(async (resp) => {
+    if (resp !== undefined) {
+      const savedAugmentObjects = get(resp, 'hits', []);
+      // gets all the saved object for this visualization
+      const savedObjectsForThisVisualization = savedAugmentObjects.filter(
+        (savedObj) => get(savedObj, 'visId', '') === AugmentVis.visId
+      );
+
+      if (maxAssociatedCount <= savedObjectsForThisVisualization.length) {
+        // eslint-disable-next-line no-throw-literal
+        throw (
+          'Cannot associate the plugin resource to the visualization due to the limit of the' +
+          'max amount of associated plugin resources (' +
+          maxAssociatedCount +
+          ') with ' +
+          savedObjectsForThisVisualization.length +
+          ' associated to the visualization'
+        );
+      }
+    }
+  });
```

Comment:
So the more I'm thinking about this, I think we will need to make this check on the server-side at the low-level api call made by the saved object client. Otherwise we still aren't covering all scenarios, and the limit could be circumvented by making alternate calls to create & save the object.

This function `createAugmentVisSavedObject()` is essentially just a helper fn / wrapper that is using the saved object loader under the hood. The loader itself is an abstraction on top of the saved object client.

To my knowledge there hasn't been set limitations on creating a saved object from the client, other than the fact that `type` and `attributes` can't be empty - see https://github.com/opensearch-project/OpenSearch-Dashboards/blob/main/src/core/public/saved_objects/saved_objects_client.ts#L244

Reply:
cc @ashwin-pc @joshuarrrr for thoughts on this.

Reply:
Maybe having this at the loader-level is still appropriate, and if users circumvent by creating these saved objs in unexpected ways (e.g., direct api calls), that is ok. Typically saved objs are quite fragile and are highly recommended not to be created manually, and altering them later on could lead to weird/unexpected behavior. I'd still like to get OSD folks opinion on this.

Reply:
Yea this is a real issue as its bypassing the setting. More enforcement would be hard too since they can bypass the normal saved object api and write to the index directly.

Reply:
If the reason to move this to the server side is to prevent the user from circumventing the limit, I'd love to know more about how risky that is. But in general it's better to put a warning near the setting than to move it to the server. Moving it to the server is more useful when we want to reduce latency or access secrets. For real safeguards, that has to be controlled on OS itself. As for saved object safeguards, yeah we don't have many today and that's a sorely needed feature.

Reply:
At the same time if you think it's necessary, feel free to add your own safeguards for your saved objects on the server. It's a judgement call

Reply:
Yea I think leaving as is would be best. Later the roadmap is that OSD would have its own data store separate from OS, so having this logic in OS would cause more problems in the future. I believe adding some documentation here would be helpful for the users.